### PR TITLE
solve the OrderedServicesCache concurrency exception

### LINKED ISSUE #11307: Investigate mvn install failure
The mvn install failure sometimes, please investigate.

 
```log
[ERROR] Tests run: 1372, Failures: 0, Errors: 2, Skipped: 0, Time elapsed: 22.192 s <<< FAILURE! - in org.apache.shardingsphere.test.integration.engine.it.dql.GeneralDQLIT
[ERROR] assertExecute[jdbc: dbtbl_with_readwrite_splitting_and_encrypt -> H2 -> Placeholder -> SELECT 1 as a](org.apache.shardingsphere.test.integration.engine.it.dql.GeneralDQLIT)  Time elapsed: 0.038 s  <<< ERROR!
java.lang.ClassCastException: org.apache.shardingsphere.authority.rule.builder.AuthorityRuleBuilder cannot be cast to org.apache.shardingsphere.infra.rule.builder.scope.SchemaRuleBuilder

[ERROR] assertExecuteQuery[jdbc: readwrite_splitting -> H2 -> Literal -> SELECT t_order.order_id,t_order.user_id,status FROM t_order t_order WHERE t_order.user_id = ? AND order_id = ?](org.apache.shardingsphere.test.integration.engine.it.dql.GeneralDQLIT)  Time elapsed: 0.035 s  <<< ERROR!
java.lang.ClassCastException: org.apache.shardingsphere.authority.rule.builder.AuthorityRuleBuilder cannot be cast to org.apache.shardingsphere.infra.rule.builder.scope.SchemaRuleBuilder
```

### PR DESCRIPTION
**Pull request by strongduanmu**
Fixes #11307.

Changes proposed in this pull request:
- solve the OrderedServicesCache concurrency exception
- use guava cache to replace ConcurrentHashMap

### CODE REVIEW

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/spi/ordered/cache/OrderedServicesCacheKey.java [hunk 1]
```diff
@@ -17,20 +17,19 @@
...
 
 /**
- * Cached ordered services.
+ * Ordered services cache key.
  */
+@EqualsAndHashCode
 @RequiredArgsConstructor
-@Getter
-public final class CachedOrderedServices {
+public class OrderedServicesCacheKey {
```
Comment (terrymanu) [inline shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/spi/ordered/cache/OrderedServicesCacheKey.java:30]: may add final with class

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/spi/ordered/cache/OrderedServicesCacheKey.java [hunk 2]
```diff
@@ -17,20 +17,19 @@
...
 import lombok.RequiredArgsConstructor;
 
 import java.util.Collection;
-import java.util.Map;
 
 /**
- * Cached ordered services.
+ * Ordered services cache key.
  */
+@EqualsAndHashCode
```
Comment (terrymanu) [inline shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/spi/ordered/cache/OrderedServicesCacheKey.java:28]: Please adjust order, constructor should the first of annotation

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/spi/ordered/OrderedSPIRegistry.java [hunk 3]
```diff
@@ -77,18 +77,18 @@
      * @param comparator comparator
      * @return registered services
      */
-    @SuppressWarnings("unchecked")
     public static <K, V extends OrderedSPI<?>> Map<K, V> getRegisteredServices(final Collection<K> types, final Class<V> orderedSPIClass, final Comparator<Integer> comparator) {
```
Comment (terrymanu) [inline shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/spi/ordered/OrderedSPIRegistry.java:80]: Is it better to adjust the param sequence too?
`orderedSPIClass` maybe the first param, `types` is second one.

**Review by terrymanu — CHANGES_REQUESTED** [review]

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/spi/ordered/cache/OrderedServicesCacheKey.java (hunk 1 above)
Comment (terrymanu) [reply shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/spi/ordered/cache/OrderedServicesCacheKey.java:30]: How about consider move this class as inner class of `OrderedServicesCache` and simplify the class name as `Key`?

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/spi/ordered/cache/OrderedServicesCache.java [hunk 4]
```diff
@@ -17,50 +17,45 @@
...
-     * @param types types
-     * @param orderedSPIClass class of ordered SPI
+     * @param cacheKey cache key
      * @param services ordered services
      * @param <K> type of key
      * @param <V> type of ordered SPI class
      */
-    public static <K, V extends OrderedSPI<?>> void cacheServices(final Collection<K> types, final Class<V> orderedSPIClass, final Map<K, V> services) {
-        CACHED_SERVICES.put(orderedSPIClass, new CachedOrderedServices(types, services));
+    public static <K, V extends OrderedSPI<?>> void cacheServices(final OrderedServicesCacheKey cacheKey, final Map<K, V> services) {
```
Comment (terrymanu) [inline shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/spi/ordered/cache/OrderedServicesCache.java:58]: Is it necessary to add generic type `K` and `V`?

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/spi/ordered/OrderedSPIRegistry.java (hunk 3 above)
Comment (strongduanmu) [reply shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/spi/ordered/OrderedSPIRegistry.java:80]: > Is it better to adjust the param sequence too?
> `orderedSPIClass` maybe the first param, `types` is second one.

@terrymanu Yes, it is better to understand. I will optimize it.

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/spi/ordered/cache/OrderedServicesCache.java (hunk 4 above)
Comment (strongduanmu) [reply shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/spi/ordered/cache/OrderedServicesCache.java:58]: > Is it necessary to add generic types `K` and `V`?

@terrymanu I will try to remove them.

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/spi/ordered/cache/OrderedServicesCacheKey.java (hunk 1 above)
Comment (strongduanmu) [reply shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/spi/ordered/cache/OrderedServicesCacheKey.java:30]: > How about consider move this class as inner class of `OrderedServicesCache` and simplify the class name as `Key`?

@terrymanu It's a goods idea, i will optimize it.

**Review by terrymanu — APPROVED** [review]

**Pull request merged**